phpBB/phpbb/storage/storage.php

```diff
@@ -18,43 +18,110 @@
  */
 class storage
 {
+	/**
+	 * @var \phpbb\storage\adapter\adapter_interface
+	 */
 	protected $adapter;
 
-	public function __construct($adapter)
+	/**
+	 * Constructor
+	 *
+	 * @param \phpbb\storage\adapter_factory	$factory
+	 * @param string							$storage_name
+	 */
+	public function __construct(adapter_factory $factory, $storage_name)
 	{
-		$this->adapter = $adapter;
+		$this->adapter = $factory->get($storage_name);
```

Comment:
you should do lazily: store the factory in a property, adds a protected get_adapter() method and everywhere in the file uses get_adapter().

get_adapter:
```
protected function get_adapter()
{
    if($this->adapter === null) {
        $this->adapter = $this->factory->get($this->storage_name);
    }

    return $this->adapter;
}
```